File: frontend/src/lib/types/trend.ts
export type TrendStatus = 'emerging' | 'active' | 'peaking' | 'declining' | 'dead';

export type SignalSource =
  | 'google_trends' | 'reddit' | 'hackernews' | 'youtube_trending'
  | 'github_trending' | 'wikipedia' | 'google_news' | 'producthunt'
  | 'npm_registry' | 'pypi_stats' | 'arxiv' | 'coingecko'
  | 'steam_charts' | 'devto' | 'lobsters' | 'stackoverflow';

export interface Trend {
  id: string;
  topic: string;
  topicSlug: string;
  description?: string;
  summary?: string;
  categoryId?: string;
  tags: string[];
  trendScore: number;
  velocityScore: number;
  correlationScore: number;
  signalStrength: number;
  sentimentScore: number;
  signalCount24h: number;
  velocity24h: number;
  acceleration: number;
  activeSources: SignalSource[];
  sourceCount: number;
  firstSource?: SignalSource;
  firstSeenAt: string;
  status: TrendStatus;
  isViral: boolean;
  isBreaking: boolean;
  viewCount: number;
  shareCount: number;
  createdAt: string;
  updatedAt: string;
}

export interface TrendFilters {
  category?: string;
  sources?: SignalSource[];
  status?: TrendStatus;
  minScore?: number;
  timeRange?: '1h' | '6h' | '24h' | '7d' | '30d';
  sortBy?: 'score' | 'velocity' | 'newest' | 'sources';
}

File: backend/app/schemas/trend.py
from pydantic import BaseModel
from datetime import datetime
from uuid import UUID
from typing import Optional


class TrendResponse(BaseModel):
    id: UUID
    topic: str
    topic_slug: str
    description: Optional[str] = None
    summary: Optional[str] = None
    category_id: Optional[UUID] = None
    status: str = "active"
    trend_score: float = 0
    velocity_24h: float = 0
    acceleration: float = 0
    signal_count_24h: int = 0
    source_count: int = 0
    active_sources: list[str] = []
    tags: list[str] = []
    first_seen_at: Optional[datetime] = None
    is_viral: bool = False
    is_breaking: bool = False
    view_count: int = 0
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class TrendListResponse(BaseModel):
    items: list[TrendResponse]
    total: int
    offset: int = 0
    limit: int = 50
